actioncable/lib/action_cable/connection/test_case.rb

```diff
@@ -0,0 +1,221 @@
+require "action_dispatch/testing/test_request"
+
+module ActionCable
+  module Connection
+    class NonInferrableConnectionError < ::StandardError
+      def initialize(name)
+        super "Unable to determine the connection to test from #{name}. " +
+          "You'll need to specify it using `tests YourConnection` in your " +
+          "test case definition."
+      end
+    end
+
+    module Assertions
+      # Asserts that the connection is rejected (via +reject_unauthorized_connection+).
+      #
+      #   # Asserts that connection without user_id fails
+      #   assert_reject_connection { connect cookies: { user_id: '' } }
+      def assert_reject_connection(&block)
+        res =
+          begin
+            block.call
+            false
+          rescue ActionCable::Connection::Authorization::UnauthorizedError
+            true
+          end
+
+        assert res, "Expected to reject connection but no rejection were made"
+      end
+    end
+
+    # We don't want to use the whole "encryption stack" for connection
+    # unit-tests, but we want to make sure that users test against the correct types
+    # of cookies (i.e. signed or encrypted or plain)
+    class TestCookieJar < ActiveSupport::HashWithIndifferentAccess
+      def signed
+        self[:signed] ||= {}.with_indifferent_access
+      end
+
+      def encrypted
+        self[:encrypted] ||= {}.with_indifferent_access
+      end
+    end
+
+    class TestRequest < ActionDispatch::TestRequest
+      attr_accessor :session, :cookie_jar
+
+      attr_writer :cookie_jar
+    end
+
+    module TestConnection
+      attr_reader :logger, :request
+
+      def initialize(request)
+        inner_logger = ActiveSupport::Logger.new(StringIO.new)
+        tagged_logging = ActiveSupport::TaggedLogging.new(inner_logger)
+        @logger = ActionCable::Connection::TaggedLoggerProxy.new(tagged_logging, tags: [])
+        @request = request
+        @env = request.env
+      end
+    end
+
+    # Superclass for Action Cable connection unit tests.
+    #
+    # == Basic example
+    #
+    # Unit tests are written as follows:
+    # 1. First, one uses the +connect+ method to simulate connection.
+    # 2. Then, one asserts whether the current state is as expected (e.g. identifiers).
+    #
+    # For example:
+    #
+    #   module ApplicationCable
+    #     class ConnectionTest < ActionCable::Connection::TestCase
+    #       def test_connects_with_cookies
+    #         # Simulate a connection
+    #         connect cookies: { user_id: users[:john].id }
+    #
+    #         # Asserts that the connection identifier is correct
+    #         assert_equal "John", connection.user.name
+    #       end
+    #
+    #       def test_does_not_connect_without_user
+    #         assert_reject_connection do
+    #           connect
+    #         end
+    #       end
+    #     end
+    #   end
+    #
+    # You can also provide additional information about underlying HTTP request:
```

Comment:
I guess it makes sense to mention the support of signed/encrypted/plain cookies, headers, env, and session in the documentation.